### ISSUE TITLE
Search box input element needs label element

### ISSUE THREAD

**Issue by jbickar** [SU-SWS/open_framework#76]
The search box input element (input#edit-keys) does not have an associated &lt;label&gt; element, and this generates accessibility errors when reviewing using achecker (http://achecker.ca/checker/index.php)

See https://commencement.stanford.edu for an example

I thought I had created an issue for this in the past but can't find it.

**Comment by jbickar** [comment]
Actually, I think it would be the input#edit-submit element

**Comment by jbickar** [comment]
This still persists, as recently as 2.06. There is no label on the input#edit-submit element, and it generates errors in AChecker.

**Comment by jbickar** [comment]
Actually, this appears to be an issue with Garland as well, so it looks like an issue with Drupal. I'll try to find the d.o issue.

**Comment by jbickar** [comment]
There seems to be some confusion as to whether this is an issue or not. I'm following up with Jennifer to learn more. http://wave.webaim.org does not report this as a problem.

If it does need a label element, here would be the code (in template.php):

function open_framework_form_alter(&$form, &$form_state, $form_id) {
  if ($form_id == 'search_block_form') {
    $form['search_block_form']['#title_display'] = 'invisible';
    $form['search_block_form']['#attributes']['class'][] = 'input-medium search-query';
    $form['search_block_form']['#attributes']['placeholder'] = t('Search this site...');
    $form['actions']['submit']['#prefix'] = '<label for "edit-submit" class="visually-hidden">Search</label>';
    $form['actions']['submit']['#attributes']['class'][] = 'btn-search';
    $form['actions']['submit']['#attributes']['alt'] = '""';
    unset($form['actions']['submit']['#value']);  
    $form['actions']['submit']['#type'] = 'image_button';
    $form['actions']['submit']['#src'] = drupal_get_path('theme', 'open_framework') . '/images/searchbutton.png';
  }
}

**Comment by hyperboy** [comment]
Thanks for looking into this.  Feel free to add $form['actions']['submit']['#prefix'] = 'Search'; to template.php and submit a pull request.

**Comment by jbickar** [comment]
Oh, it stripped out my html code. Let's see if I can get this right with Markdown:

```
function open_framework_form_alter(&$form, &$form_state, $form_id) {
  if ($form_id == 'search_block_form') {
    $form['search_block_form']['#title_display'] = 'invisible';
    $form['search_block_form']['#attributes']['class'][] = 'input-medium search-query';
    $form['search_block_form']['#attributes']['placeholder'] = t('Search this site...');
    $form['actions']['submit']['#prefix'] = '<label for "edit-submit" class="visually-hidden">Search</label>';
    $form['actions']['submit']['#attributes']['class'][] = 'btn-search';
    $form['actions']['submit']['#attributes']['alt'] = '""';
    unset($form['actions']['submit']['#value']);    
    $form['actions']['submit']['#type'] = 'image_button';
    $form['actions']['submit']['#src'] = drupal_get_path('theme', 'open_framework') . '/images/searchbutton.png';
  }
}
```

**Comment by jbickar** [comment]
The above code is still a little funky with the alt attribute on the searchbutton.png, so I need to get to the bottom of that. It outputs it as alt="""". If I use:
    $form['actions']['submit']['#attributes']['alt'] = "";

Then it just outputs "alt" with no "=" or value. Jennifer says it needs a null alt, and not to repeat the value in the label in the alt attribute for the label.

**Comment by jbickar** [comment]
Per http://www.jimthatcher.com/webcourse8.htm and discussion on the WebAIM mailing list, image buttons must have both a "src" and an "alt" attribute.

Therefore the implementation in Open Framework is correct.

```
<input type="image" src="https://example.stanford.edu/sites/all/themes/open_framework/images/searchbutton.png" name="submit" id="edit-submit" alt="Search" class="btn-search form-submit">
```
